aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -150,6 +154,116 @@ struct GroupRead<T: Transaction> {
+pub enum DelayedFieldRead {
+    // Represents a full read - that value has been returned to the caller,
+    // meaning that read is valid only if value is identical.
+    Value {
+        value: DelayedFieldValue,
+    },
+    // Represents a restricted read - where a range of values that satisfy DeltaHistory
+    // are all valid and produce the same outcome.
+    // Only boolean outcomes of "try_add_delta" operations have been returned to the caller,
+    // and so we need to respect that those return the same outcome when doing the validation.
+    // Running inner_value is kept only for internal bookeeping - and is used to as a value against
+    // which results are computed, but is not checked for read validation.
+    HistoryBounded {
+        restriction: DeltaHistory,
+        max_value: u128,
+        inner_value: DelayedFieldValue,
+    },
+}
+
+impl DelayedFieldRead {
+    fn get_kind(&self) -> DelayedFieldReadKind {
+        use DelayedFieldRead::*;
+        match self {
+            Value { .. } => DelayedFieldReadKind::Value,
+            HistoryBounded { .. } => DelayedFieldReadKind::HistoryBounded,
+        }
+    }
+
+    /// If the reads contains sufficient information, return it, otherwise return None
+    pub(crate) fn filter_by_kind(
+        &self,
+        min_kind: DelayedFieldReadKind,
+    ) -> Option<DelayedFieldRead> {
+        let self_kind = self.get_kind();
+        if self_kind >= min_kind {
+            Some(self.clone())
+        } else {
+            None
+        }
+    }
+
+    // A convenience method, since the same key can be read in different modes, producing
+    // different DataRead / ReadKinds. Returns true if self has >= kind than other, i.e.
+    // contains more or equal information, and is consistent with the information in other.
+    fn contains(&self, other: &DelayedFieldRead) -> DataReadComparison {
+        use DelayedFieldRead::*;
+        match (&self, other) {
+            (Value { value: v1, .. }, Value { value: v2, .. }) => {
+                if v1 == v2 {
+                    DataReadComparison::Contains
+                } else {
+                    DataReadComparison::Inconsistent
+                }
+            },
+            (
+                HistoryBounded {
+                    restriction: h1,
+                    inner_value: v1,
+                    max_value: m1,
+                },
+                HistoryBounded {
+                    restriction: h2,
+                    inner_value: v2,
+                    max_value: m2,
+                },
+            ) => {
+                // TODO see if we want to perform checks on max value/delta
+                if v1 == v2 && m1 == m2 && h1.stricter_than(h2) {
+                    DataReadComparison::Contains
+                } else {
+                    DataReadComparison::Inconsistent
+                }
+            },
+            (HistoryBounded { .. }, Value { .. }) => DataReadComparison::Insufficient,
+            (
```

Comment:
Is `contains(self, other) = true` when every possible value that satisfies `other` also satisfies `self`?
In that case, the current code for `(HistoryBounded { .. }, Value { .. })` and `(Value { .. }, HistoryBounded { .. })` conditions may need to be swapped.

Reply:
Returns true if self has >= kind than other, i.e.
contains more or equal information, and is consistent with the information in other.

maybe renaming to stricter_than is clearer, though @georgemitenkov seems to not like that name :)